README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ivdoctr

<!-- badges: start -->

<!-- badges: end -->

Instrumental variables are a useful tool in causal inference. In order
to be valid, researchers impose both formal beliefs (instrumental
relevance and the exclusion restriction) and informal beliefs (e.g.,
correlation between endogenous treatment and the error term). The goal
of `ivdoctr` is to formalize those beliefs and quantifies how sensitive
a researcher’s instrumental variables are to measurement error and
instrument endogeneity. Using data and researcher’s beliefs on
measurement error and instrument endogeneity, this package generates the
space of consistent beliefs across measurement error, instrument
endogeneity, and instrumental relevance for IV regressions.

## Installation

You can install the released version of ivdoctr from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("ivdoctr")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("fditraglia/ivdoctr")
```

## Example

This section illustrates how to use the `ivdoctr` package in practice.
This example comes from “The colonial origins of comparative
development: An empirical investigation” by Acemoglu, Johnson, and
Robinson (2001). The authors study the effect of institutions on GDP per
capita across 64 countries. Since institutional quality is endogenous,
they use differences in mortality rates of early western settlers across
colonies as an instrumental variable. The regression specification is as
follows:

![](https://render.githubusercontent.com/render/math?math=%5Cbegin%7Balign*%7D%20%5Clog\(GDP%20%2F%20capita\)%20%26%3D%20%5Calpha_0%20%2B%20%5Cbeta%20Institutions%20%2B%20u%20%20%5C%5C%20Institutions%20%26%3D%20%5Calpha_1%20%2B%20%5Cpi%20%5Clog\(Settler%20Mortality\)%20%2B%20v%20%5Cend%7Balign*%7D)

The authors state that there is likely a positive correlation between
institutional quality and the error term, which could come from reverse
causality (e.g., wealthier societies can afford better institutions) or
omitted variables (e.g., rule of law or British culture are positively
correlated with present-day institutional quality). This positive
correlation is a researcher belief that can be input into `ivdoctr`
using the `r_TstarU_restriction` argument that accepts a 2-column matrix
of bounds. For the exercise, we use 0.9 as the conservative upper bound
on the extent of the endogeneity.

The authors also state that up to 40% of the measure of “institutions”
is noise. Measurement error is
![1-](https://render.githubusercontent.com/render/math?math=1-%5Ckappa),
so
![\[0.6, 1\]](https://render.githubusercontent.com/render/math?math=%5Ckappa%20%5Cin%20%5B0.6%2C%201%5D)
is the translation of this belief. The code below implements these
beliefs and runs the estimation and saves the TeX table to
“colonial.tex”:

``` r
library(ivdoctr)
endog <- matrix(c(0, 0.9), nrow = 1)
meas <- matrix(c(0.6, 1), nrow = 1)

colonial_example <- makeExample(y_name = "logpgp95", T_name = "avexpr", 
                                z_name = "logem4", data = colonial,
                                controls = NULL, robust = FALSE,
                                r_TstarU_restriction = endog,
                                k_restriction = meas,
                                example_name = "Colonial Origins")
                                
makeTable("colonial.tex", binary = FALSE, colonial_example)
```

To explore the surface of estimates consistent with the researcher’s
beliefs, `ivdoctr` also generates an interactive 3D plot of the surface:

``` r
library(ivdoctr)
endog <- matrix(c(0, 0.9), nrow = 1)
meas <- matrix(c(0.6, 1), nrow = 1)

plot_3d_beta(y_name = "logpgp95", T_name = "avexpr", 
             z_name = "logem4", data = colonial, 
             r_TstarU_restriction = endog,
             k_restriction = meas)
```

## Usage

This package exports three main functions:

  - `makeExample()`: Generates the TeX code that can be used to build a
    regression summary table.
  - `makeTable()`: Generates the TeX code for a stand-alone regression
    table and saves it to the specified file. Uses the output of
    `makeExample()` as its input.
  - `plot_3d_beta()`: Generates an interactive 3D plot illustrating the
    relationship between the causal estimates, instrument endogeneity,
    instrument invalidity, and measurement error.

Both `makeExample` and `plot_3d_beta` use the same primary inputs. Users
input the name of the dataset (`data`), the name of the dependent
variable (`y_name`), the name of the treatment variable(s) (`T_name`),
the name(s) of the instrument(s) (`z_name`), and the names of the
control variables (`controls`). Without any additional arguments, the
functions will output the identified set. If users have beliefs over
measurement error and/or instrument endogeneity, they can specify those
using `k_restriction` and `r_TstarU_restriction`, respectively.


NEWS.md
# ivdoctr 1.0

+ Initial CRAN release


R/paperTable.R
# #----------------------- Colonial Origins -------------------------------------
# load("./data/colonial.rda")
#
# # Part I of table
# summary_1 <- get_estimates("logpgp95", "avexpr", "logem4", colonial)
# obs <- get_observables("logpgp95", "avexpr", "logem4", colonial)
# summary_2 <- get_bounds_unrest(obs)
# stats <- list(n = summary_1$n,
#               b_OLS = summary_1$b_OLS,
#               se_OLS = summary_1$se_OLS,
#               b_IV = summary_1$b_IV,
#               se_IV = summary_1$se_IV,
#               k_lower = summary_2$k$Lower,
#               r_uz_bound = ifelse(abs(summary_2$r_uz$Lower) == 1,
#                                   summary_2$r_uz$Upper,
#                                   summary_2$r_uz$Lower))
# make_I(stats, "Colonial Origins")
#
# # Part II of table
# ## First prior
# bounds_1 <- draw_bounds("logpgp95", "avexpr", "logem4", colonial, controls = NULL,
#                       r_TstarU_restriction = c(0, 0.9),
#                       k_restriction = c(0.00001, 0.6),
#                       n_draws = 5000, Jeffreys = TRUE)
# freq_1 <- summarize_bounds(bounds_1)
#
# posterior_1 <- draw_posterior("logpgp95", "avexpr", "logem4", colonial, controls = NULL,
#                              r_TstarU_restriction = c(0, 0.9),
#                             k_restriction = c(0.000001, 0.6),
#                             n_RF_draws = 1000,
#                           n_IS_draws = 1000,
#                           Jeffreys = TRUE,
#                           resample = FALSE)
# bayes_1 <- summarize_posterior(posterior_1)
#
# stats <- list(p_empty = freq_1$p_empty,
#               p_valid = freq_1$p_valid,
#               beta_lower_median = freq_1$restricted$median,
#               beta_upper_median = freq_1$restricted$median,
#               beta_lower_lower_bound = freq_1$restricted$lower[1],
#               beta_lower_upper_bound = freq_1$restricted$upper[1],
#               beta_upper_lower_bound = freq_1$restricted$lower[2],
#               beta_upper_upper_bound = freq_1$restricted$upper[2],
#               r_uz_median = bayes_1$HPDI$median[1],
#               beta_bayes_median = bayes_1$HPDI$median[2],
#               r_uz_lower_bound = bayes_1$HPDI$lower[1],
#               beta_bayes_lower_bound = bayes_1$HPDI$lower[2],
#               r_uz_upper_bound = bayes_1$HPDI$upper[1],
#               beta_bayes_upper_bound = bayes_1$HPDI$upper[2])
# make_II_III(stats, "$(\\kappa, \\rho_{T^*u}) \\in (0, 0.6] \\times [0, 0.9]")
#
# ## Second prior
# bounds_2 <- draw_bounds("logpgp95", "avexpr", "logem4", colonial, controls = NULL,
#                         r_TstarU_restriction = c(0, 0.9),
#                         k_restriction = c(0.6, 1),
#                         n_draws = 5000, Jeffreys = TRUE)
# freq_2 <- summarize_bounds(bounds_2)
#
# posterior_2 <- draw_posterior("logpgp95", "avexpr", "logem4", colonial, controls = NULL,
#                               r_TstarU_restriction = c(0, 0.9),
#                               k_restriction = c(0.6, 1),
#                               n_RF_draws = 1000,
#                               n_IS_draws = 1000,
#                               Jeffreys = TRUE,
#                               resample = FALSE)
# bayes_2 <- summarize_posterior(posterior_2)
#
# stats <- list(p_empty = freq_2$p_empty,
#               p_valid = freq_2$p_valid,
#               beta_lower_median = freq_2$restricted$median,
#               beta_upper_median = freq_2$restricted$median,
#               beta_lower_lower_bound = freq_2$restricted$lower[1],
#               beta_lower_upper_bound = freq_2$restricted$upper[1],
#               beta_upper_lower_bound = freq_2$restricted$lower[2],
#               beta_upper_upper_bound = freq_2$restricted$upper[2],
#               r_uz_median = bayes_2$HPDI$median[1],
#               beta_bayes_median = bayes_2$HPDI$median[2],
#               r_uz_lower_bound = bayes_2$HPDI$lower[1],
#               beta_bayes_lower_bound = bayes_2$HPDI$lower[2],
#               r_uz_upper_bound = bayes_2$HPDI$upper[1],
#               beta_bayes_upper_bound = bayes_2$HPDI$upper[2])
# make_II_III(stats, "$(\\kappa, \\rho_{T^*u}) \\in (0.6, 1] \\times [0, 0.9]")
#
#
# y_name <- "logpgp95"
# T_name <- "avexpr"
# z_name <- "logem4"
# data <- colonial
# controls <- NULL
# r_TstarU_restriction <- c(0, 0.9)
# k_restriction <- c(0.00001, 0.6)
# n_RF_draws <- 1000
# n_IS_draws <- 1000
# Jeffreys <- TRUE
# resample <- FALSE
#
# # r_TstarU_lower <- r_TstarU_min[!empty]
# # r_TstarU_upper <- r_TstarU_max[!empty]
# # k_lower <- k_min[!empty]
# # k_upper <- k_max[!empty]
# # obs <- obs_draws[!empty, ]
#
# ## Was Weber Wrong?
# load("./data/weber.rda")
# controls <- c("f_young", "f_jew", "f_fem", "f_ortsgeb", "f_pruss", "hhsize",
#               "lnpop", "gpop", "f_miss")
# summary_1 <- get_estimates("f_rw", "f_prot", "kmwittenberg", controls = controls, weber)
# obs <- get_observables("f_rw", "f_prot", "kmwittenberg", weber, controls = controls)
# summary_2 <- get_bounds_unrest(obs)
# stats <- list(n = summary_1$n,
#               b_OLS = summary_1$b_OLS,
#               se_OLS = summary_1$se_OLS,
#               b_IV = summary_1$b_IV,
#               se_IV = summary_1$se_IV,
#               k_lower = summary_2$k$Lower,
#               r_uz_bound = ifelse(abs(summary_2$r_uz$Lower) == 1,
#                                   summary_2$r_uz$Upper,
#                                   summary_2$r_uz$Lower))
# # make_I(stats, "Was Weber Wrong?")
